flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sources/BoundedTableSource.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.sources;
+
+import org.apache.flink.annotation.Experimental;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+
+/**
+ * Defines an external bounded table and provides access to its data.
+ *
+ * @param <T> Type of the bounded {@link DataStream} created by this {@link TableSource}.
+ */
+@Experimental
+public interface BoundedTableSource<T> extends TableSource<T> {
+
+	/**
+	 * Returns the data of the table as a {@link DataStream}.
+	 *
+	 *<p>NOTE: This method is for internal use only for defining a {@link TableSource}.
+	 *       Do not use it in Table API programs.
+	 */
+	DataStream<T> getBoundedDataStream(StreamExecutionEnvironment execEnv);
```

Comment:
I thought we would like to uncouple the `BoundedTableSource` from DataStream API and let them return `InputFormat`? This would allow us to add the source also in `flink-table-common`.

Reply:
As I said in the description, we also want to reuse existing streaming source implementation to support bounded source, for example: bounded Kafka source. If we just returns `InputFormat`, we have to implement duplicate code for lots of connectors. 

From my point of view, no matter returns `InputFormat` or `DataStream`, it is a temporary solution which will be refactored/deprecated when we introduce unified table source.

Reply:
The difference between this solution and the discussed solution is that we cannot provide backwards compatibility in future versions. Once the new source interfaces are ready we can and need to provide InputFormat-to-SourceFunction and OutputFormat-to-SinkFunction wrappers for supporting old connectors of the DataSet API. However, if we give users access to `StreamExecutionEnvironment`, it is difficult to have an upgrade story. It exposes too much of the underlying layers.

Another problem is that we cannot support batch connectors in the legacy planner for the next version. Which further complicates the feature-to-planner compatibility list.

Reply:
We considered it again, and found only some message queue sources are using `SourceFunction`. Common batch data sources are all using `InputFormat`. So I think we can go back to the discussed solution, returns an `InputFormat`.  And provide a unified table source in 1.10 to support bounded message queue.

Reply:
This sounds good, in order for both runners to support the same batch sources I think we pretty much have to support `InputFormat`/`OutputFormat` because these are the common denominator. 👌

Reply:
Thanks @aljoscha @twalthr, I have a discussion with Timo in slack and know the new idea about `BoundedTableSource`. I pushed a new commit for it. You can take a look if it matches what you think.